src/value_generators/src/DocumentGenerator.cpp

```diff
@@ -937,6 +938,139 @@ class ObjectIdGenerator : public Generator<bsoncxx::types::b_oid> {
+    }
+}
+
+/**
+ * Validate that the uuid sub type is correct.
+ * @param sub_type
+ *   the binary sub type value
+ * @raises InvalidValueGeneratorSyntax if the sub type is not correct
+*  Currntly only supports sub type 4.
+ */
+void validateUuidSubType(bsoncxx::binary_sub_type sub_type) {
+
+    if (bsoncxx::binary_sub_type::k_uuid != sub_type) {
+        std::stringstream msg;
+        msg << "Invalid binary sub_type. UUID only supports bsoncxx::binary_sub_type::k_uuid("
+            << static_cast<int64_t>(bsoncxx::binary_sub_type::k_uuid) << "), got " << static_cast<int64_t>(sub_type);
+        BOOST_LOG_TRIVIAL(warning) << " UuidGenerator " << msg.str();
+        BOOST_THROW_EXCEPTION(InvalidValueGeneratorSyntax(msg.str()));
+    }
+}
+
+/**
+ * Convert ascii hex nibble to binary.
+ * @param c
+ *   the ascii hex nibble
+ * @return
+ *   the binary version of the hex nibble
+ */
+uint8_t hexNibbleToBin(uint8_t c) {
+    if (c >= '0' && c <= '9') {
+        c -= '0';
+    } else  if (c >= 'a' && c <= 'f') {
+        c = c - 'a' + 10;
+    } else  if (c >= 'A' && c <= 'F') {
+        c = c - 'A' + 10;
+    }
+    return c;
+}
+
+/**
+ * Convert hex string Uuid to binary.
+ * @param hex
+ *   the input hex string, '-' will be ignored.
+ * @param uuid
+ *   the array to store the uuid.
+ * @return the output array
+ */
+uint8_t* hex2BinUuid(std::string hex, uint8_t* uuid) {
+    hex.erase(std::remove_if(hex.begin(), hex.end(),
+                             [&](char c) { return c == '-'; } ),
+              hex.end());
+
+    for(std::string::iterator it = hex.begin(); it != hex.end(); ++it) {
+        int pos = std::distance(hex.begin(), it) / 2;
+        uuid[pos] = (hexNibbleToBin(*it) << 4) + hexNibbleToBin(*++it);
+    }
+    return uuid;
+}
+
+/**
+ * `{^UUID: {hex: "3b241101-e2bb-4255-8caf-4136c566a962"} }`
+ *  Generator for non-legacy UUIDs (0x04). hex can itself be a generator so the string can be
+ *  generated in many fashions ^Join, ^FormatString, ... etc.
+ *
+ *  The input format accepted for the hex field is 8-4-4-4-12 format. The code is agnostic as to the
+ *  UUID version.
+ *
+ *  see https://en.wikipedia.org/wiki/Universally_unique_identifier
+ */
+class UuidGenerator : public Generator<bsoncxx::types::b_binary> {
+public:
+    UuidGenerator(const Node& node, GeneratorArgs generatorArgs)
+        : _rng{generatorArgs.rng},
+          _node{node},
+          _hexGen{stringGenerator(node["hex"], generatorArgs)},
+          // If there is no subType then use bsoncxx::binary_sub_type::k_uuid
+          _subTypeGen{node["subType"]
+                          ? intGenerator(node["subType"], generatorArgs)
+                          : std::make_unique<ConstantAppender<int64_t>>(
+                                static_cast<int64_t>(bsoncxx::binary_sub_type::k_uuid))} {}
+
+    bsoncxx::types::b_binary evaluate() override {
+        auto hex = _hexGen->evaluate();
```

Comment:
I'm not sure we want to validate each and every time. I see why you want to do so. What is the overhead from the call?